src/lib/storage/chunk_encoder.cpp

```diff
@@ -12,41 +12,114 @@
+                                                          const SegmentEncodingSpec& encoding_spec) {
+  std::shared_ptr<BaseSegment> result;
+  resolve_data_type(data_type, [&](auto type) {
+    using ColumnDataType = typename decltype(type)::type;
+
+    // TODO(anyone): After #1489, build segment statistics in encode_segment() instead of encode_chunk()
+    // and store them within the segment instead of a chunk-owned list of statistics.
+
+    if (const auto reference_segment = std::dynamic_pointer_cast<const ReferenceSegment>(segment)) {
+      Fail("Reference segments cannot be encoded. Try to pass the reference segment(s).");
+    }
+
+    // Check if early exit is possible when passed segment is currently unencoded and the same is requested.
+    const auto unencoded_segment = std::dynamic_pointer_cast<const ValueSegment<ColumnDataType>>(segment);
+    if (unencoded_segment && encoding_spec.encoding_type == EncodingType::Unencoded) {
+      result = segment;
+      return;
+    }
+
+    // Check for early exit, when requested segment encoding is already being used for the passed segment.
+    const auto encoded_segment = std::dynamic_pointer_cast<const BaseEncodedSegment>(segment);
+    if (encoded_segment && encoded_segment->encoding_type() == encoding_spec.encoding_type) {
+      // Encoded segments do not need to be reencoded when the requested encoding is already present and either
+      // (i) no vector compression is given or (ii) the vector encoding is also already present. Hence, with
+      // {EncodingType::Dictionary} being in place, {EncodingType::Dictionary, VectorCompressionType::SimdBp128}
+      // does not reencode the segment.
+      if (!encoding_spec.vector_compression_type ||
+          (encoding_spec.vector_compression_type && encoded_segment->compressed_vector_type() &&
+           encoding_spec.vector_compression_type ==
+               parent_vector_compression_type(*encoded_segment->compressed_vector_type()))) {
+        result = segment;
+        return;
+      }
+    }
+
+    // In case of unencoded re-encoding, an any segment iterable is used to manually setup
+    // the data vectors for a ValueSegment. If another encoding is requested, the segment
+    // encoding utitilies are used (which create and call the according encoder).
+    if (encoding_spec.encoding_type == EncodingType::Unencoded) {
+      pmr_concurrent_vector<ColumnDataType> values;
+      pmr_concurrent_vector<bool> null_values;
+
+      auto iterable = create_any_segment_iterable<ColumnDataType>(*segment);
+      iterable.with_iterators([&](auto it, auto end) {
+        const auto segment_size = std::distance(it, end);
+        values.reserve(segment_size);
+        null_values.reserve(segment_size);
+
+        for (; it != end; ++it) {
+          const auto segment_item = *it;
+          const auto is_null = segment_item.is_null();
+          null_values.push_back(is_null);
```

Comment:
Why do you use reserve+push_back if you know the final size?

Reply:
Would you use `reserve()` + `[i]`? I thought we measured that both are equally fast? If not, I'll change it.

Reply:
That would be UB. `resize() + []` however, is significantly faster, because it does neither capacity checks nor has to update the actual size of the vector.

Reply:
Sorry, that's what I meant. We need an additional run variable though. Will change.

Reply:
Changed.